src/db.rs

```diff
@@ -1286,8 +1298,21 @@ impl Db {
     /// ```
     pub async fn flush(&self) -> Result<(), SlateDBError> {
         if self.inner.wal_enabled() {
-            self.inner.flush_wals().await
+            // DBInner::flush() will frozen the active WAL and flush all imm_wals to memtable
+            self.inner.flush().await?;
+            {
+                // TODO(asukamilet): refactor flush logic
+                let mut guard = self.inner.state.write();
+                let wal_id = guard.last_written_wal_id() + 1;
+                guard.freeze_memtable(wal_id)?;
+            }
+            self.inner.flush_memtables().await
         } else {
+            {
+                let mut guard = self.inner.state.write();
+                let wal_id = guard.last_written_wal_id() + 1;
```

Comment:
this should be `guard.last_written_wal_id()`. When WAL is off, all memtable flushes use current last-written wal id.